### PR TITLE
small snakemake pipeline for cross validation

### PR DESCRIPTION
**Pull request by silastittes**
Had to make a few tweaks to how the data files were read in. 
Added dependencies I needed to conda env, but happy to remove if we don't want them for end users. 

If you like the snakemake set up, should be easy to build ontop of this. 

Also easy to run in HPC setting using a cluster config. For example, 

I have the following config on talpas at  `~/.config/snakemake/talapas/config.yaml`

```
cluster:
        mkdir -p logs/{rule} &&
        sbatch
                --partition={resources.partition}
                --account=kernlab
                --cpus-per-task={resources.threads}
                --mem={resources.mem_mb}
                --time={resources.time_min}
                --job-name=smk-{rule}-{wildcards}
                --output=logs/{rule}/{rule}-{wildcards}-%j.out
                --error=logs/{rule}/{rule}-{wildcards}-%j.err
                --gres=gpu:{resources.gpu}
default-resources:
        - partition=kern
        - time_min=60
        - mem_mb=5000
        - threads=1
        - gpu=0
restart-times: 3
max-jobs-per-second: 10
max-status-checks-per-second: 1
local-cores: 1
latency-wait: 120
jobs: 200
keep-going: True
rerun-incomplete: True
printshellcmds: True
scheduler: greedy
use-conda: True
```

Then to send each command as batch jobs I'd use:

`snakemake --profile --profile ~/.config/snakemake/talapas/ -C script=crossvalidation.py basedir=simulation/density_saddle/out_2877096093782_stats`

### CODE REVIEW

File: crossvalidation.py [hunk 1]
```diff
@@ -47,12 +47,13 @@
 results_file = os.path.join(outdir, 'results.pkl')
 
 # load spatial and genetic data
-spatial_data = pd.read_csv(os.path.join(outdir, params['spatial_data']), index_col=0).rename(
-        columns={"site_name": "name", "long": "x", "lat": "y"}
+spatial_data = pd.read_csv(params['spatial_data']).rename(
+        columns={"name": "name", "long": "x", "lat": "y"}
```
Comment (petrelharp) [inline crossvalidation.py:51]: Okay, so with this change the working directory of this script is now assumed to be the directory the json file lives in, right? I don't like this so much because 
- paths in the json file should be relative to where that file lives; otherwise they don't make sense;
- and so if you run `crossvalidation path/to/json` then you'll only get what you expect if you're actually in the same directory as the json

File: example_usage.ipynb [hunk 2]
```diff
@@ -38,10 +38,21 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
```
Comment (petrelharp) [inline example_usage.ipynb:41]: Did you actually change anything in this notebook? If not, could you revert the changes to it? (damn notebooks and git)

File: simulation/simulation_workflow.ipynb [hunk 3]
```diff
@@ -16,10 +16,32 @@
     "from PIL import Image\n",
     "from scipy import interpolate\n",
     "import matplotlib.pyplot as plt\n",
-    "\n",
```
Comment (petrelharp) [inline simulation/simulation_workflow.ipynb]: how about this one - is it changed?

File: crossvalidation.py (hunk 1 above)
Comment (silastittes) [reply crossvalidation.py:51]: Gotcha! I was having trouble getting it to execute, but I think `index_col=0` was the problem, not the path join stuff. Will revert.

File: example_usage.ipynb (hunk 2 above)
Comment (silastittes) [reply example_usage.ipynb:41]: Not intentionally! Will revert!

File: simulation/simulation_workflow.ipynb (hunk 3 above)
Comment (silastittes) [reply simulation/simulation_workflow.ipynb]: I did add a cell or two to run SLiM and `compute_stats.py` -- mostly in efforts to get my head around stuff. happy to revert.

File: crossvalidation.py (hunk 1 above)
Comment (silastittes) [reply crossvalidation.py:51]: I think I must be setting up the `crossvalidation.json` for `setup_crossval_grid.py`. 

The only example input on my end in the repo is `data/nebria/crossvalidation.json`

Without the edits here, if I run

```
python setup_crossval_grid.py data/nebria/crossvalidation.json 3 0.1 100 2 1 10
python crossvalidation.py data/nebria/xval_0/xval_params.json
```
I get `FileNotFoundError: [Errno 2] No such file or directory: 'data/nebria/xval_0/stats.csv'`

Comment (silastittes) [reply crossvalidation.py:51]: Tried a few more things and read the code more. *If* I'm reading correctly, `outdir` is the path to one of the `xval_*` directories that `setup_crossval_grid.py` makes, but then here it tries to look for the data in those paths and can't find them. The change I made here assumed the .json file passed to `setup_crossval_grid.py` has the relative path to the input files. At least when I tried it with these edits I was getting:

`FileNotFoundError: [Errno 2] No such file or directory: 'data/nebria/xval_0/stats.csv'`

Comment (petrelharp) [reply crossvalidation.py:51]: Ah-ha, that's true; So, the problem there is elsewhere, including with `setup_crossval_grid.py`. So, it should always be true that if you're in the same directory as the json file, that the paths it refers to exist. So, for instance, `simulation/bias_saddle/crossvalidation.json` has
```
    "spatial_data": "stats.csv",
    "genetic_data": "pairstats.csv",
```
... but however 
```
$ ls simulation/bias_saddle/{stats.csv,pairstats.csv}
ls: cannot access 'simulation/bias_saddle/stats.csv': No such file or directory
ls: cannot access 'simulation/bias_saddle/pairstats.csv': No such file or directory
```
Ah-ha, that's one problem. Notice that we get the error running even `python3 crossvalidation.py simulation/bias_saddle/crossvalidation.json`

Actually I've got the data files in a subdirectory:
```
$ ls simulation/bias_saddle/out_2874405695390_stats
params.json  rep371579.ibd.png  rep371579.pairstats.csv  rep371579.stats.csv
```
and so in fact I should have `simulation/bias_saddle/crossvalidation.json` like
```
    "spatial_data": "out_2874405695390_stats/rep371579.stats.csv",
    "genetic_data": "out_2874405695390_stats/rep371579.pairstats.csv",
```
However, there's still a problem then with `setup_crossval_grid.py` - it ought to modify the paths so that they're still valid, since the new json files they set up are in subdirectories.

I'll fix these things up.

**Comment by petrelharp** [comment]
I cherry-picked the snakefile and things over on `main/`; if you'd like to add the other changes carry on.

**Comment by silastittes** [comment]
Gonna close as you got the useful bit incorporated already.
